atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java

```diff
@@ -1810,11 +1849,33 @@ private void ensureNoEmptyValues(Map<Cell, byte[]> values) {
         }
     }
 
-    private void putWritesAndLogIfTooLarge(Map<Cell, byte[]> values, SortedMap<Cell, byte[]> writes) {
+    private void putWritesAndLogIfTooLarge(
+            Map<Cell, byte[]> values,
+            ConcurrentNavigableMap<Cell, byte[]> writes,
+            Map<Cell, ChangeMetadata> metadata,
+            ConcurrentNavigableMap<Cell, ChangeMetadata> metadataForWrites) {
         for (Map.Entry<Cell, byte[]> e : values.entrySet()) {
             byte[] val = MoreObjects.firstNonNull(e.getValue(), PtBytes.EMPTY_BYTE_ARRAY);
             Cell cell = e.getKey();
-            if (writes.put(cell, val) == null) {
+            AtomicBoolean wasNull = new AtomicBoolean(false);
+            boolean hasMetadata = metadata.containsKey(cell);
+            writes.compute(cell, (k, oldVal) -> {
+                wasNull.set(oldVal == null);
+                try {
+                    Thread.sleep(new Random().nextInt(2));
+                } catch (InterruptedException ex) {
+                    throw new RuntimeException(ex);
+                }
+                // If we are not writing metadata for a value, we have to remove any previously stored metadata since
+                // it may not be valid for the new value.
+                if (hasMetadata) {
+                    metadataForWrites.put(cell, metadata.get(cell));
+                } else {
+                    metadataForWrites.remove(cell);
+                }
+                return val;
+            });
+            if (wasNull.get()) {
```

Comment:
This is the spicy bit. Previously, we could just do `map.put()` and get the old value as a result (which we need to do some counting). This refactor preserves this behavior, but allows us to also update the metadata map in an atomic fashion. Note that the mapping function provided in `compute` may be run multiple times. This is ok because the semantics is still the same: The call that is last to place a value in `writes` also places the corresponding value in `metadataForWrites`, which is the desired behavior.